Import co-locates double record units for relationships

tinwelint opened the pull request:
If it sees that the number of relationships will likely be big enough that many of the relationships will require secondary record units.

This PR removes some home brew id sequences and generators that the importer previously had in favor of the ones in the stores. The importer looks at the estimates and decides to allocate two ids per relationships if number of relationships are large enough and the format is the high limit format. This to reduce contention in the import and faster reads from the db after import.

tinwelint committed:
Allocate two relationship record units next to each other

if number of estimated relationships exceeds some magic threshold

tinwelint committed:
Lets linking stages also use the prepare id sequence

please squasch with previous commit at some point

tinwelint committed:
Simplified PropertyEncoderStep by removing lots of code

due to an id issue where more ids than expected were actually used

tinwelint committed:
Considers double record units in disk estimation

tinwelint committed:
Better memory usage reporting

tinwelint committed:
Added testing about allocating double record units in import

tinwelint committed:
Fixes an issue updating unused records in importer

tinwelint committed:
Proper input data in test

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/store/SecondaryUnitPrepareIdSequenceTest.java:
```diff
@@ -0,0 +1,81 @@
+        try
+        {
+            prepared.nextId();
+            fail( "Shoul've failed" );
```
`Shoud`

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/ImportLogic.java:
```diff
@@ -157,16 +157,35 @@ public void initialize( Input input ) throws IOException
-        dependencies.satisfyDependencies( input.calculateEstimates( neoStore.getPropertyStore().newValueEncodedSizeCalculator() ),
-                idMapper, neoStore, nodeRelationshipCache );
+        Estimates inputEstimates = input.calculateEstimates( neoStore.getPropertyStore().newValueEncodedSizeCalculator() );
+        sanityCheckEstimatesWithRecordFormat( inputEstimates );
```
would really like to see tests for those sanity checks

tinwelint replied:
Pulled it out and unit tested that, looks better?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/store/BatchingNeoStores.java:
```diff
@@ -77,6 +80,8 @@
 public class BatchingNeoStores implements AutoCloseable, MemoryStatsVisitor.Visitable
 {
+    // Empirical and slightly defensive threshold where relationship records seem to start requiring double record units.
+    private static final long DOUBLE_RELATIONSHIP_RECORD_UNIT_THRESHOLD = 1L << 33;
```
while this empirical metric can be good enough, i think we need to add a bit of description how it was built (based on what assumptions) or direct future us into "requiredDataLength" from record formats etc.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/store/BatchingNeoStores.java:
```diff
@@ -418,4 +425,18 @@ public void success()
+    {
+        doubleRelationshipRecordUnits =
+                // TODO figure out in a good way if this is the high limit format
+                recordFormats.relationship().getMaxId() > 2L << 36 &&
```
2L << 36 should be extracted as constant

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/store/BatchingNeoStores.java:
```diff
@@ -418,4 +425,18 @@ public void success()
+    public boolean determineDoubleRelationshipRecordUnits( Estimates inputEstimates )
+    {
+        doubleRelationshipRecordUnits =
+                // TODO figure out in a good way if this is the high limit format
```
and probably TODO can be removed

MishaDemianenko reviewed (changes requested):
Couple of small comments, good otherwise

tinwelint committed:
Pull out and test estimate vs maxIds sanity checking

tinwelint committed:
More testing on determining double rel record units

Also introducing SECONDARY_RECORD_UNITS capability to easily determine whether or not
the format supports secondary units instead of guessing by looking at maxId

tinwelint commented:
@MishaDemianenko updated
